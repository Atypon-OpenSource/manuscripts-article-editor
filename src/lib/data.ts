/*!
 * Â© 2019 Atypon Systems LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { UserProfileWithAvatar } from '@manuscripts/manuscript-editor'
import { UserProfile } from '@manuscripts/manuscripts-json-schema'
import { RxDocument } from 'rxdb'

export const buildUser = async (
  doc: RxDocument<UserProfile>
): Promise<UserProfileWithAvatar> => {
  const item = doc.toJSON() as UserProfileWithAvatar

  const attachment = await doc.getAttachment('image')

  if (attachment) {
    item.avatar = window.URL.createObjectURL(await attachment.getData())
  }

  return item
}

export const buildUserMap = async (docs: Array<RxDocument<UserProfile>>) => {
  const output: Map<string, UserProfileWithAvatar> = new Map()

  for (const doc of docs) {
    const item = await buildUser(doc)

    output.set(item.userID, item)
  }

  return output
}
