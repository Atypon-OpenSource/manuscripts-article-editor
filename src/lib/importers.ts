import JSZip from 'jszip'
import { extname } from 'path'
import { FIGURE } from '../transformer/object-types'
import {
  Attachments,
  ContainedModel,
  Figure,
  Model,
  ModelAttachment,
} from '../types/models'
import { generateAttachmentFilename } from './exporter'
import { convert } from './pressroom'

type Importer = (file: File) => Promise<Model[]>

interface Importers {
  [key: string]: Importer
}

export interface JsonModel
  extends ContainedModel,
    Attachments,
    ModelAttachment {
  _id: string
  bundled: boolean
  collection: string
  contentType: string
}

export interface ProjectDump {
  version: string
  data: JsonModel[]
}

const modelHasObjectType = <T extends Model>(objectType: string) => (
  model: Model
): model is T => {
  return model.objectType === objectType
}

export const readManuscriptFromBundle = async (
  zip: JSZip
): Promise<ProjectDump> => {
  const json = await zip.file('index.manuscript-json').async('text')

  return JSON.parse(json)
}

const importProjectBundle = async (result: Blob) => {
  const zip = await new JSZip().loadAsync(result)

  const doc = await readManuscriptFromBundle(zip)

  if (doc.version !== '2.0') {
    throw new Error(`Unsupported version: ${doc.version}`)
  }

  // TODO: validate?
  // TODO: ensure default data is added

  const items = doc.data
    .filter(item => !item.bundled)
    .filter(item => item._id)
    .map(item => {
      delete item._rev
      delete item.collection

      return item
    })

  const folder = zip.folder('Data')

  await Promise.all(
    items
      .filter(modelHasObjectType<Figure>(FIGURE))
      .map(async (item: Figure) => {
        const filename = generateAttachmentFilename(item._id)

        try {
          item.attachment = {
            id: filename, // TODO: original name?
            type: item.contentType || 'image/png',
            data: await folder.file(filename).async('blob'),
          }
        } catch (error) {
          // tslint:disable-next-line:no-console
          console.error(`Could not retrieve attachment from zip: ${error}`)
          // continue without attachment
        }
      })
  )

  return items
}

const importConvertedFile: Importer = async (file: File) => {
  const form = new FormData()
  form.append('file', file)

  const result = await convert(form, '.manuproj')

  // download(result, 'manuscript.zip')

  return importProjectBundle(result)
}

const importers: Importers = {
  '.docx': importConvertedFile,
  '.html': importConvertedFile,
  '.md': importConvertedFile,
  '.manuproj': importProjectBundle,
}

export const openFilePicker = (): Promise<File> =>
  new Promise((resolve, reject) => {
    const input = document.createElement('input')
    input.type = 'file'
    input.accept = Object.keys(importers).join(',')
    input.addEventListener('change', () => {
      if (input.files && input.files.length) {
        resolve(input.files[0])
      } else {
        reject(new Error('No file was received'))
      }
    })
    input.click()
  })

export const importFile = async (file: File) => {
  const extension = extname(file.name)

  return importers[extension](file)
}
