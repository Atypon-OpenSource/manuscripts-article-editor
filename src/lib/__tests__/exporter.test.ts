jest.mock('../pressroom')

import JSZip from 'jszip'
import data from '../__fixtures__/example-manuscript.json'
import {
  exportProject,
  generateAttachmentFilename,
  generateDownloadFilename,
} from '../exporter'
import { readManuscriptFromBundle } from '../importers'
import { buildComponentMap } from './util'

describe('exporter', () => {
  test('generates a filename for an attachment', () => {
    const result = generateAttachmentFilename('MPFigure:1234-ABCD-EFGH-1234')
    expect(result).toBe('MPFigure_1234-ABCD-EFGH-1234')
  })

  test('generates a filename for a manuscript title', () => {
    const result = generateDownloadFilename('An example manuscript')
    expect(result).toBe('AnExampleManuscript')
  })

  test('generates a filename for a manuscript title with markup', () => {
    const result = generateDownloadFilename('An <b>example</b> manuscript')
    expect(result).toBe('AnExampleManuscript')
  })

  test('exports a manuscript', async () => {
    const componentMap = buildComponentMap(data)

    // `result` is the blob that would be sent for conversion, echoed back
    const result = await exportProject(componentMap, 'docx')
    expect(result).toBeInstanceOf(Blob)
    expect(result.size).toBe(17002)

    const zip = await new JSZip().loadAsync(result)
    const manuscript = await readManuscriptFromBundle(zip)

    expect(manuscript.version).toBe('2.0')
    expect(manuscript.data).toHaveLength(33)
    expect(manuscript).toMatchSnapshot('exported-manuscript')
  })
})
