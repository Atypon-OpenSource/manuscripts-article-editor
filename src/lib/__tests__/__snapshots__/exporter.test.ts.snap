// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`exporter exports a manuscript: exported-manuscript 1`] = `
Object {
  "data": Array [
    Object {
      "_id": "MPTable:E6862029-A3E8-4510-AC46-D7B278F8BD1F",
      "_rev": "4-bc49f73d29f26806d2bb80b070a11303",
      "caption": "Caption",
      "collection": "tables",
      "contents": "<table class=\\"MPElement MPTableStyle_E497A750-2E15-44E2-8414-60A95890031D MPParagraphStyle_86F12F93-F46C-449D-A951-BEE88A69537B\\" id=\\"MPTableElement:5B51441E-128C-4110-D4F7-D3437096371E\\" data-contained-object-id=\\"MPTable:E6862029-A3E8-4510-AC46-D7B278F8BD1F\\"><thead><tr><th class=\\"placeholder\\" data-placeholder-text=\\"Header 1\\"></th><th class=\\"placeholder\\" data-placeholder-text=\\"Header 2\\"></th></tr></thead><tbody><tr><td class=\\"\\" data-placeholder-text=\\"Data\\"><b>dfgsdfgsdfg</b></td><td class=\\"placeholder\\" data-placeholder-text=\\"Data\\"></td></tr><tr><td class=\\"\\" data-placeholder-text=\\"Data\\"><i>sdfgsdfgsdfg</i></td><td class=\\"placeholder\\" data-placeholder-text=\\"Data\\"></td></tr></tbody><tfoot><tr><td class=\\"placeholder\\" data-placeholder-text=\\"Footer 1\\"></td><td class=\\"placeholder\\" data-placeholder-text=\\"Footer 2\\"></td></tr></tfoot></table>",
      "createdAt": 1438718342,
      "objectType": "MPTable",
      "sessionID": "a15897d3-b99a-431f-e9e1-4bae9f14411d",
    },
    Object {
      "_id": "MPManuscript:D16458EF-E9CF-409D-A2B2-77D08272B62C",
      "_rev": "5-e265f6402bb31850412f969b5b4ea6d5",
      "collection": "Manuscripts",
      "createdAt": 1438715371.977981,
      "manuscriptType": 1,
      "objectType": "MPManuscript",
      "sessionID": "ab33960d-4e0d-4b01-dc6a-c0e8249a711b",
      "title": "Essay regarding the greatness of Steve Ballmer.",
      "updatedAt": 1438715413.195572,
    },
    Object {
      "TeXRepresentation": "< Equation >",
      "_id": "MPEquation:3CE14D2F-E9E6-45BE-BF38-B98687FB8A9E",
      "_rev": "1-cee928562abcc64ce1b7592ab8f991ba",
      "collection": "equations",
      "createdAt": 1438720739,
      "objectType": "MPEquation",
      "sessionID": "45c1c9aa-531e-406a-daaa-ae0a688683f9",
      "title": "Equation",
    },
    Object {
      "_id": "MPHighlight:1af53ab6-514b-4063-f74e-704dfcf5d4e4",
      "_rev": "1-eb4b800bab66cdc7491ed5f69aef513c",
      "createdAt": 1438733779.610575,
      "objectType": "MPHighlight",
      "quote": "little",
      "ranges": Array [
        Object {
          "end": "//*[@id='MPParagraphElement:6804DA8C-B8A3-48ED-9E09-CF4AC931C4A8']",
          "endOffset": 27,
          "start": "//*[@id='MPParagraphElement:6804DA8C-B8A3-48ED-9E09-CF4AC931C4A8']",
          "startOffset": 21,
        },
      ],
      "text": "sdfgsdfgsdfg",
      "updatedAt": 1438733779.610575,
    },
    Object {
      "_id": "MPHighlight:2881dd99-d9b6-4e70-a952-a17434676ad0",
      "_rev": "1-8071a0727e21d004ac5441414b2b0b54",
      "createdAt": 1438799390.308151,
      "objectType": "MPHighlight",
      "quote": ".",
      "ranges": Array [
        Object {
          "end": "//*[@id='MPParagraphElement:6804DA8C-B8A3-48ED-9E09-CF4AC931C4A8']",
          "endOffset": 57,
          "start": "//*[@id='MPParagraphElement:6804DA8C-B8A3-48ED-9E09-CF4AC931C4A8']",
          "startOffset": 56,
        },
      ],
      "text": "sdfgsdfgsdfg",
      "updatedAt": 1438799390.308151,
    },
    Object {
      "_id": "MPHighlight:39ec0d9e-590b-4af7-f369-f896be46716c",
      "_rev": "1-dd9be99fd4bda677f20263b0daa032c6",
      "createdAt": 1438796681.044155,
      "objectType": "MPHighlight",
      "quote": "sdfg",
      "ranges": Array [
        Object {
          "end": "//*[@id='MPParagraphElement:214D7C10-A78D-4511-C022-46AF6C9425E9']",
          "endOffset": 4,
          "start": "//*[@id='MPParagraphElement:214D7C10-A78D-4511-C022-46AF6C9425E9']",
          "startOffset": 0,
        },
      ],
      "text": "sdfgsdfgsdfgsdfgsdf sdfg sdfg sdfg sdfg sdfg sdfg sdfg sdfg sdf sdfg sdf gsdf gsdf gsdf gsdf gsdfg sdf gsdf gsdfg sdfg sdfg sdf gsdfg sdfg sdfg sdfg sdfg sdfg sdf sdfg sdfg sdf sdfg ",
      "updatedAt": 1438796681.044155,
    },
    Object {
      "_id": "MPHighlight:3d67515f-feb7-4c1b-f33f-54497dfe85d5",
      "_rev": "1-89364524dffc7f638fe54b30dbb6e202",
      "createdAt": 1438798254.218312,
      "objectType": "MPHighlight",
      "quote": "sdfg",
      "ranges": Array [
        Object {
          "end": "//*[@id='MPParagraphElement:D910B981-0E9D-4BDF-D9DA-D6654D1A0A3B']",
          "endOffset": 4,
          "start": "//*[@id='MPParagraphElement:D910B981-0E9D-4BDF-D9DA-D6654D1A0A3B']",
          "startOffset": 0,
        },
      ],
      "text": "sdfgsdfsdfg",
      "updatedAt": 1438798254.218312,
    },
    Object {
      "_id": "MPHighlight:496d03d4-14df-49ff-9442-eaf4cf0e7fc2",
      "_rev": "1-92a12742a8ec80b0cf0427f6a425ced6",
      "createdAt": 1438797450.704594,
      "objectType": "MPHighlight",
      "quote": "sdfg",
      "ranges": Array [
        Object {
          "end": "//*[@id='MPParagraphElement:14FDC9F9-C7E9-44F8-A8A3-D1F24C1AF7A0']",
          "endOffset": 4,
          "start": "//*[@id='MPParagraphElement:14FDC9F9-C7E9-44F8-A8A3-D1F24C1AF7A0']",
          "startOffset": 0,
        },
      ],
      "text": "sdfg",
      "updatedAt": 1438797450.704594,
    },
    Object {
      "_id": "MPHighlight:57fd9c52-3d75-44fb-c446-102c9ce079ee",
      "_rev": "1-014e90afd5d8720d7da3ce82a6945299",
      "createdAt": 1438797557.110174,
      "objectType": "MPHighlight",
      "quote": "sdfg",
      "ranges": Array [
        Object {
          "end": "//*[@id='MPParagraphElement:6804DA8C-B8A3-48ED-9E09-CF4AC931C4A8']",
          "endOffset": 24,
          "start": "//*[@id='MPParagraphElement:6804DA8C-B8A3-48ED-9E09-CF4AC931C4A8']",
          "startOffset": 20,
        },
      ],
      "text": "sdfgsdfgsdfg",
      "updatedAt": 1438797557.110174,
    },
    Object {
      "_id": "MPHighlight:5976c382-4152-4664-e4cb-63ef0a8f1190",
      "_rev": "1-8938dc782d3011d7ecbabfbc7612510b",
      "createdAt": 1438797446.542481,
      "objectType": "MPHighlight",
      "quote": "sdfg",
      "ranges": Array [
        Object {
          "end": "//*[@id='MPParagraphElement:D910B981-0E9D-4BDF-D9DA-D6654D1A0A3B']",
          "endOffset": 4,
          "start": "//*[@id='MPParagraphElement:D910B981-0E9D-4BDF-D9DA-D6654D1A0A3B']",
          "startOffset": 0,
        },
      ],
      "text": "sdgfsdfg",
      "updatedAt": 1438797446.542481,
    },
    Object {
      "_id": "MPHighlight:5d309325-4484-419e-9896-d7592b86a112",
      "_rev": "1-d24764bfee32da36ebcb6a5514e22795",
      "createdAt": 1438798362.705209,
      "objectType": "MPHighlight",
      "quote": "sdfg",
      "ranges": Array [
        Object {
          "end": "//*[@id='MPParagraphElement:B60BB38F-5F26-444C-865D-BE3741CCFF17']",
          "endOffset": 12,
          "start": "//*[@id='MPParagraphElement:B60BB38F-5F26-444C-865D-BE3741CCFF17']",
          "startOffset": 8,
        },
      ],
      "text": "dsfgsdfg",
      "updatedAt": 1438798362.705209,
    },
    Object {
      "_id": "MPHighlight:659d7a78-df5a-457d-fbe3-a094c121502a",
      "_rev": "1-82e95973c6e194ebc67b764de8f6710e",
      "createdAt": 1438795863.445524,
      "objectType": "MPHighlight",
      "quote": "asdf",
      "ranges": Array [
        Object {
          "end": "//*[@id='MPParagraphElement:6804DA8C-B8A3-48ED-9E09-CF4AC931C4A8']",
          "endOffset": 13,
          "start": "//*[@id='MPParagraphElement:6804DA8C-B8A3-48ED-9E09-CF4AC931C4A8']",
          "startOffset": 9,
        },
      ],
      "text": "sdfgsdfgsdfg",
      "updatedAt": 1438795863.445524,
    },
    Object {
      "_id": "MPHighlight:7917b0ca-1aa4-4dc9-9e75-d7ef0f973166",
      "_rev": "1-6415604f5ff700d53caa2aa6986ef43f",
      "createdAt": 1438733767.337305,
      "objectType": "MPHighlight",
      "quote": "fads",
      "ranges": Array [
        Object {
          "end": "//*[@id='MPParagraphElement:6804DA8C-B8A3-48ED-9E09-CF4AC931C4A8']",
          "endOffset": 32,
          "start": "//*[@id='MPParagraphElement:6804DA8C-B8A3-48ED-9E09-CF4AC931C4A8']",
          "startOffset": 28,
        },
      ],
      "text": "asdfasdfasdf",
      "updatedAt": 1438733767.337305,
    },
    Object {
      "_id": "MPHighlight:92003126-6ba6-4819-f635-687bca2e9e2b",
      "_rev": "2-334b91725c804cee1f0ce090ecda3fc1",
      "createdAt": 1438792241.306557,
      "objectType": "MPHighlight",
      "quote": "",
      "ranges": Array [],
      "text": "sdfgsdfgsdfg",
      "updatedAt": 1438795123.108963,
    },
    Object {
      "_id": "MPHighlight:9fbe9c31-6e15-4767-998c-70a3835d94e8",
      "_rev": "1-847aa27b7663352fdd6124fe25fb3a6f",
      "createdAt": 1438799017.931055,
      "objectType": "MPHighlight",
      "quote": "s df",
      "ranges": Array [
        Object {
          "end": "//*[@id='MPParagraphElement:8C924F7A-1EFD-4B40-BFE4-1C10AFFDE372']",
          "endOffset": 9,
          "start": "//*[@id='MPParagraphElement:8C924F7A-1EFD-4B40-BFE4-1C10AFFDE372']",
          "startOffset": 5,
        },
      ],
      "text": "sdfgsdfgsdfg",
      "updatedAt": 1438799017.931055,
    },
    Object {
      "_id": "MPHighlight:a9f3cf90-96eb-4c58-9006-a445cafc88fc",
      "_rev": "1-1cf9bf7408d4fbc275ce3d8317bfd30f",
      "createdAt": 1438799023.074736,
      "objectType": "MPHighlight",
      "quote": "df",
      "ranges": Array [
        Object {
          "end": "//*[@id='MPParagraphElement:8C924F7A-1EFD-4B40-BFE4-1C10AFFDE372']",
          "endOffset": 13,
          "start": "//*[@id='MPParagraphElement:8C924F7A-1EFD-4B40-BFE4-1C10AFFDE372']",
          "startOffset": 11,
        },
      ],
      "text": "",
      "updatedAt": 1438799023.074736,
    },
    Object {
      "_id": "MPHighlight:ad8e3fd2-d905-4b06-f2a9-5a14b9aa35f0",
      "_rev": "1-395349573500deda41581739361a0beb",
      "createdAt": 1438797569.597403,
      "objectType": "MPHighlight",
      "quote": "g",
      "ranges": Array [
        Object {
          "end": "//*[@id='MPParagraphElement:6804DA8C-B8A3-48ED-9E09-CF4AC931C4A8']",
          "endOffset": 41,
          "start": "//*[@id='MPParagraphElement:6804DA8C-B8A3-48ED-9E09-CF4AC931C4A8']",
          "startOffset": 40,
        },
      ],
      "text": "sdfgsdfgsdfgsdfg",
      "updatedAt": 1438797569.597403,
    },
    Object {
      "_id": "MPHighlight:c67c8dd8-3f8b-44d8-c345-d021c78450e1",
      "_rev": "1-3ea6926720bfba8ac6d31320796ca65d",
      "createdAt": 1438795847.182957,
      "objectType": "MPHighlight",
      "quote": "little",
      "ranges": Array [
        Object {
          "end": "//*[@id='MPParagraphElement:6804DA8C-B8A3-48ED-9E09-CF4AC931C4A8']",
          "endOffset": 27,
          "start": "//*[@id='MPParagraphElement:6804DA8C-B8A3-48ED-9E09-CF4AC931C4A8']",
          "startOffset": 21,
        },
      ],
      "text": "sdfgsdfgsdfg",
      "updatedAt": 1438795847.182957,
    },
    Object {
      "_id": "MPHighlight:de8b5e51-ad3b-4d6e-ab89-0171d50824b4",
      "_rev": "1-d82be91e8e11390755e4d0fbffc3a05e",
      "createdAt": 1438795831.977257,
      "objectType": "MPHighlight",
      "quote": "really",
      "ranges": Array [
        Object {
          "end": "//*[@id='MPParagraphElement:6804DA8C-B8A3-48ED-9E09-CF4AC931C4A8']",
          "endOffset": 15,
          "start": "//*[@id='MPParagraphElement:6804DA8C-B8A3-48ED-9E09-CF4AC931C4A8']",
          "startOffset": 9,
        },
      ],
      "text": "sdfgsdfgsdfg",
      "updatedAt": 1438795831.977257,
    },
    Object {
      "_id": "MPHighlight:e1fbee74-8a9f-4314-967c-7e2c59e08eda",
      "_rev": "1-96af09ab2a56bbb759f0588a19e4233e",
      "createdAt": 1438797563.84876,
      "objectType": "MPHighlight",
      "quote": "sdfg sdfg sdfg",
      "ranges": Array [
        Object {
          "end": "//*[@id='MPParagraphElement:6804DA8C-B8A3-48ED-9E09-CF4AC931C4A8']",
          "endOffset": 39,
          "start": "//*[@id='MPParagraphElement:6804DA8C-B8A3-48ED-9E09-CF4AC931C4A8']",
          "startOffset": 25,
        },
      ],
      "text": "",
      "updatedAt": 1438797563.84876,
    },
    Object {
      "_id": "MPHighlight:f5e6f37d-7f1b-4dd6-c174-06448f2d8288",
      "_rev": "1-81bf08332b859139e6746dc8734bb33c",
      "createdAt": 1438795834.763864,
      "objectType": "MPHighlight",
      "quote": "sdfg",
      "ranges": Array [
        Object {
          "end": "//*[@id='MPParagraphElement:6804DA8C-B8A3-48ED-9E09-CF4AC931C4A8']",
          "endOffset": 20,
          "start": "//*[@id='MPParagraphElement:6804DA8C-B8A3-48ED-9E09-CF4AC931C4A8']",
          "startOffset": 16,
        },
      ],
      "text": "",
      "updatedAt": 1438795834.763864,
    },
    Object {
      "_id": "MPHighlight:f81a374f-4956-4881-b1ef-4827a597135c",
      "_rev": "1-2ee4bb3110058d505cdadc9beed88fcb",
      "createdAt": 1438799551.198289,
      "objectType": "MPHighlight",
      "quote": "sdfg",
      "ranges": Array [
        Object {
          "end": "//*[@id='MPParagraphElement:D3BC4DBA-14CF-4232-BE7F-413E158B9A0A']",
          "endOffset": 16,
          "start": "//*[@id='MPParagraphElement:D3BC4DBA-14CF-4232-BE7F-413E158B9A0A']",
          "startOffset": 12,
        },
      ],
      "text": "sdfgsdfgsdfg",
      "updatedAt": 1438799551.198289,
    },
    Object {
      "_id": "MPFigureElement:58837A7A-9C62-4459-FAE5-6425BC7676DC",
      "_rev": "10-f1dede8b6c243f359ff80bb364bd27ef",
      "caption": "",
      "collection": "elements",
      "containedObjectIDs": Array [
        "MPFigure:D3A86C07-4316-4425-B837-AAAF438345B1",
        "MPFigure:9B62EE51-6867-4E39-89F9-DB8D1E9492A3",
      ],
      "createdAt": 1438715407,
      "elementType": "figure",
      "figureLayout": "MPFigureLayout:F7F4C16D-B63A-4181-BC96-44C097DDD810",
      "figureStyle": "MPFigureStyle:319B1B50-5DF3-44FB-9E47-941286B80364",
      "objectType": "MPFigureElement",
      "sessionID": "3478833b-b9bd-4601-c184-984cb8b8a12a",
      "updatedAt": 1438720127.46186,
    },
    Object {
      "_id": "MPParagraphElement:059840E8-0E29-4F96-8990-C327A1BB1007",
      "_rev": "2-72448cb4dab1be95330a2fffeed1c434",
      "collection": "elements",
      "contents": "<p id=\\"MPParagraphElement:059840E8-0E29-4F96-8990-C327A1BB1007\\" class=\\"MPElement MPParagraphStyle_86F12F93-F46C-449D-A951-BEE88A69537B\\" data-object-type=\\"MPParagraphElement\\">asdfasdfasdf</p>",
      "createdAt": 1438802517,
      "elementType": "p",
      "objectType": "MPParagraphElement",
      "paragraphStyle": "MPParagraphStyle:86F12F93-F46C-449D-A951-BEE88A69537B",
      "sessionID": "3478833b-b9bd-4601-c184-984cb8b8a12a",
    },
    Object {
      "_id": "MPParagraphElement:65EDABB4-A8F6-41A7-AF2E-2977854942CB",
      "_rev": "2-2abdff9dc640b4bd7ce9e9b21a28097d",
      "collection": "elements",
      "contents": "<p id=\\"MPParagraphElement:65EDABB4-A8F6-41A7-AF2E-2977854942CB\\" class=\\"MPElement MPParagraphStyle_86F12F93-F46C-449D-A951-BEE88A69537B\\" data-object-type=\\"MPParagraphElement\\"><span class=\\"annotator-hl annotator-hl-temporary\\" data-annotation-id=\\"MPHighlight:85ad9842-e413-4c59-858a-4877bdde21d9\\">sdfgsdfgsdfgsdfg</span></p>",
      "createdAt": 1438798936,
      "elementType": "p",
      "objectType": "MPParagraphElement",
      "paragraphStyle": "MPParagraphStyle:86F12F93-F46C-449D-A951-BEE88A69537B",
      "sessionID": "cd2549ac-afcf-4f41-f111-63918e593351",
    },
    Object {
      "_id": "MPParagraphElement:6B4AEB63-0DD6-4263-A542-AEBF9E81B7F1",
      "_rev": "2-e6ede1aa5f890f67a43dd1e9a75f12c6",
      "collection": "elements",
      "contents": "<p id=\\"MPParagraphElement:6B4AEB63-0DD6-4263-A542-AEBF9E81B7F1\\" class=\\"MPElement MPParagraphStyle_86F12F93-F46C-449D-A951-BEE88A69537B\\" data-object-type=\\"MPParagraphElement\\">sdfgsdfgsdf sdfg sdfg sdfg sdfg sdfg sdfg sfgsfgsdfgsdf sdfg sdfgs dfgsd fgsdfgsdfg sdfgs dfsgsdfg sdfg sdfg sdfg sdfg sdfg sdfg sdfgsdfg sdfgsdfgs dfgsdfgsdfgsdfgsdfg sdfg sdfg sdfg sdfg sdfgsdfgsdfgsdfgsdfg</p>",
      "createdAt": 1438799323,
      "elementType": "p",
      "objectType": "MPParagraphElement",
      "paragraphStyle": "MPParagraphStyle:86F12F93-F46C-449D-A951-BEE88A69537B",
      "sessionID": "5c3a8358-5c1f-45e9-8e01-64b5124c3f08",
    },
    Object {
      "_id": "MPParagraphElement:70D712CF-C2B1-46F1-B36A-A059D0D1981A",
      "_rev": "2-c4f200a331adaa09a271636261446948",
      "collection": "elements",
      "contents": "<p id=\\"MPParagraphElement:70D712CF-C2B1-46F1-B36A-A059D0D1981A\\" class=\\"MPElement MPParagraphStyle_86F12F93-F46C-449D-A951-BEE88A69537B\\" data-object-type=\\"MPParagraphElement\\">safasdfasdf</p>",
      "createdAt": 1438802515,
      "elementType": "p",
      "objectType": "MPParagraphElement",
      "paragraphStyle": "MPParagraphStyle:86F12F93-F46C-449D-A951-BEE88A69537B",
      "sessionID": "3478833b-b9bd-4601-c184-984cb8b8a12a",
    },
    Object {
      "_id": "MPParagraphElement:7A46815B-C6D8-4565-9F1B-CA16EDFEDB6C",
      "_rev": "2-8450bc71554e2ffcbfc954dcf2ccbbe0",
      "collection": "elements",
      "contents": "<p id=\\"MPParagraphElement:7A46815B-C6D8-4565-9F1B-CA16EDFEDB6C\\" class=\\"MPElement MPParagraphStyle_86F12F93-F46C-449D-A951-BEE88A69537B\\" data-object-type=\\"MPParagraphElement\\">asdfasdfasdf</p>",
      "createdAt": 1438802516,
      "elementType": "p",
      "objectType": "MPParagraphElement",
      "paragraphStyle": "MPParagraphStyle:86F12F93-F46C-449D-A951-BEE88A69537B",
      "sessionID": "3478833b-b9bd-4601-c184-984cb8b8a12a",
    },
    Object {
      "_id": "MPParagraphElement:7B5DDE01-2822-4E42-A8B1-3BC9D982B035",
      "_rev": "3-6a265ad9864011f263a2f8d0993ba2b9",
      "collection": "elements",
      "contents": "<p id=\\"MPParagraphElement:7B5DDE01-2822-4E42-A8B1-3BC9D982B035\\" class=\\"MPElement MPParagraphStyle_86F12F93-F46C-449D-A951-BEE88A69537B\\" data-object-type=\\"MPParagraphElement\\">asdfasdf asdfasdf <span class=\\"annotator-hl annotator-hl-temporary\\" data-annotation-id=\\"MPHighlight:ef45dba6-e149-425c-eb7c-88a9f5babc7c\\">asdfa</span> sdfa sdfasd faasdfasd asdf</p>",
      "createdAt": 1438799672,
      "elementType": "p",
      "objectType": "MPParagraphElement",
      "paragraphStyle": "MPParagraphStyle:86F12F93-F46C-449D-A951-BEE88A69537B",
      "sessionID": "c14610d2-2d21-442b-8e4f-488e4fc96318",
    },
    Object {
      "_id": "MPParagraphElement:8C924F7A-1EFD-4B40-BFE4-1C10AFFDE372",
      "_rev": "6-af0f36a024064dc25efa59c7a10de4b8",
      "collection": "elements",
      "contents": "<p id=\\"MPParagraphElement:8C924F7A-1EFD-4B40-BFE4-1C10AFFDE372\\" class=\\"MPElement MPParagraphStyle_86F12F93-F46C-449D-A951-BEE88A69537B\\" data-object-type=\\"MPParagraphElement\\">sdfgs<span class=\\"annotator-hl\\" data-annotation-id=\\"MPHighlight:9fbe9c31-6e15-4767-998c-70a3835d94e8\\">s </span><span class=\\"annotator-hl\\" data-annotation-id=\\"MPHighlight:9fbe9c31-6e15-4767-998c-70a3835d94e8\\">df</span>gs<span class=\\"annotator-hl\\" data-annotation-id=\\"MPHighlight:a9f3cf90-96eb-4c58-9006-a445cafc88fc\\">df</span>g</p>",
      "createdAt": 1438798996,
      "elementType": "p",
      "objectType": "MPParagraphElement",
      "paragraphStyle": "MPParagraphStyle:86F12F93-F46C-449D-A951-BEE88A69537B",
      "sessionID": "c46b3ef1-3cf9-4d5a-da80-c7162d0900e0",
    },
    Object {
      "_id": "MPParagraphElement:D3BC4DBA-14CF-4232-BE7F-413E158B9A0A",
      "_rev": "3-5ecb81e03fc8c3a6ea4014451cb67fd9",
      "collection": "elements",
      "contents": "<p id=\\"MPParagraphElement:D3BC4DBA-14CF-4232-BE7F-413E158B9A0A\\" class=\\"MPElement MPParagraphStyle_86F12F93-F46C-449D-A951-BEE88A69537B\\" data-object-type=\\"MPParagraphElement\\">sdfgsdfgsdf <span class=\\"annotator-hl\\" data-annotation-id=\\"MPHighlight:f81a374f-4956-4881-b1ef-4827a597135c\\">sdfg</span> sdfg sdfg sdfg</p>",
      "createdAt": 1438799545,
      "elementType": "p",
      "objectType": "MPParagraphElement",
      "paragraphStyle": "MPParagraphStyle:86F12F93-F46C-449D-A951-BEE88A69537B",
      "sessionID": "c9f5ec1e-b0e2-48a9-9a73-ac3606ea1061",
    },
    Object {
      "_id": "MPParagraphElement:D4164064-66B5-4D2E-8C77-DAA8BAFF68E2",
      "_rev": "3-4897660f3d2d80b802f5eba3a7284fe2",
      "collection": "elements",
      "contents": "<p id=\\"MPParagraphElement:D4164064-66B5-4D2E-8C77-DAA8BAFF68E2\\" class=\\"MPElement MPParagraphStyle_86F12F93-F46C-449D-A951-BEE88A69537B\\" data-object-type=\\"MPParagraphElement\\">sdgfsdfgsdfgsdfgsdfg</p>",
      "createdAt": 1438799475,
      "elementType": "p",
      "objectType": "MPParagraphElement",
      "paragraphStyle": "MPParagraphStyle:86F12F93-F46C-449D-A951-BEE88A69537B",
      "sessionID": "e654b00f-9e86-49e7-e0cb-c822d1bf674f",
    },
    Object {
      "_id": "MPParagraphElement:F4A379D9-352F-4BD5-FCC5-EC6EB217D0A7",
      "_rev": "2-aca884879a79c604c909703b048dd18e",
      "collection": "elements",
      "contents": "<p id=\\"MPParagraphElement:F4A379D9-352F-4BD5-FCC5-EC6EB217D0A7\\" class=\\"MPElement MPParagraphStyle_86F12F93-F46C-449D-A951-BEE88A69537B\\" data-object-type=\\"MPParagraphElement\\"><span class=\\"annotator-hl annotator-hl-temporary\\" data-annotation-id=\\"MPHighlight:fe78ae05-4431-4734-c48e-4a61099cf062\\">sdfgsdfgsdfgsdfgsdf</span> sdfg sdfg</p>",
      "createdAt": 1438799428,
      "elementType": "p",
      "objectType": "MPParagraphElement",
      "paragraphStyle": "MPParagraphStyle:86F12F93-F46C-449D-A951-BEE88A69537B",
      "sessionID": "e255093b-2271-4bbc-9732-8a02ae440ca8",
    },
  ],
  "version": "2.0",
}
`;
